igor_src/Utils.cpp

```diff
@@ -466,3 +466,53 @@ uint64_t draw_random_64bits_seed(){
 	}
 	return random_seed;
 }
+
+
+UMCodonTable CodonTableStandard = {
+	{"TTT", "F"}, {"TTC", "F"}, {"TTA", "L"}, {"TTG", "L"}, 
+	{"TCT", "S"}, {"TCC", "S"}, {"TCA", "S"}, {"TCG", "S"}, 
+	{"TAT", "Y"}, {"TAC", "Y"}, {"TGT", "C"}, {"TGC", "C"}, 
+	{"TGG", "W"}, {"CTT", "L"}, {"CTC", "L"}, {"CTA", "L"}, 
+	{"CTG", "L"}, {"CCT", "P"}, {"CCC", "P"}, {"CCA", "P"}, 
+	{"CCG", "P"}, {"CAT", "H"}, {"CAC", "H"}, {"CAA", "Q"}, 
+	{"CAG", "Q"}, {"CGT", "R"}, {"CGC", "R"}, {"CGA", "R"}, 
+	{"CGG", "R"}, {"ATT", "I"}, {"ATC", "I"}, {"ATA", "I"}, 
+	{"ATG", "M"}, {"ACT", "T"}, {"ACC", "T"}, {"ACA", "T"}, 
+	{"ACG", "T"}, {"AAT", "N"}, {"AAC", "N"}, {"AAA", "K"}, 
+	{"AAG", "K"}, {"AGT", "S"}, {"AGC", "S"}, {"AGA", "R"}, 
+	{"AGG", "R"}, {"GTT", "V"}, {"GTC", "V"}, {"GTA", "V"}, 
+	{"GTG", "V"}, {"GCT", "A"}, {"GCC", "A"}, {"GCA", "A"}, 
+	{"GCG", "A"}, {"GAT", "D"}, {"GAC", "D"}, {"GAA", "E"}, 
+	{"GAG", "E"}, {"GGT", "G"}, {"GGC", "G"}, {"GGA", "G"}, 
+	{"GGG", "G"},
+	// Stop  codon
+	{"TAA", "*"}, {"TAG", "*"}, {"TGA", "*"},
+	// Start codon
+	{"TTG", "s"}, {"CTG", "s"}, {"ATG", "s"}
+};
+
+
+/*
+ * \param seq: DNA sequence to translate.
+ * \return amino acid sequence. 
+ */
+string translate(string seq){
+	size_t codonPos = 0;
+	const size_t codonLen = 3;
+	size_t seqLen = seq.length();
+	string strCodon = "";
+	string AA ="";
+	string AAChain ="";
+	bool stopCodonQ = false;
+	while ((codonPos < seqLen) and !stopCodonQ){
+		strCodon = seq.substr(codonPos, codonLen);
+//		cout << strCodon << " codonPos: " << codonPos << endl;
+		codonPos = codonPos + codonLen;
+		AA = CodonTableStandard[strCodon];
+		if (AA == "*"){
+			stopCodonQ = true;
+		}
+		AAChain = AAChain + AA; //.second;			
+	}
+	return AAChain;
```

Comment:
Well, I don't know if there is a standard or used symbol to denote an incomplete codon.
So, in that case, the function will not be considered codon fragments.
For instance, for a zero modulo 3 we get something like
TGTGCCTGG  ==> CAW
and
TGTGCCTG ==> CA
TGTGCCT ==> CA
Biopython and OLGA do the same thing. I'm also not considering cases where a Valine is incomplete for instance. I saw some software that completes the amino acid and do something like this
TGTGCCGTA ==> CAV
TGTGCCGT ==> CAV
Because CG[A, C, T or G ]  ==> V
and I personally don't like that. However, if you find a better solution to that let me know ;)

Reply:
Well I fully agree with you about this valine stuff, but I might be even more conservative on this!
From what I understand the CDR3 only makes sense when defined by the conserved Cysteine/Tryptophane/Phenylalanin, if the CDR3 length is not 0 modulo 3 then this definition doesn't hold anymore (because there's a high chance that the conserved residue is gone and probably the whole structure of the immune receptor with it).
This is fine as long as the user is aware of it and checks against the NT sequence and/or inframe boolean that the amino acid sequence makes sense in downstream analysis. But frankly I don't think we should rely on this and we should anticipate possible mistakes by specifying extra NTs e.g with '-'. Your examples would become:
TGTGCCTGG ==> CAW
and
TGTGCCTG ==> CA-
TGTGCCT ==> CA-
I think this is much clearer for the user and much less error prone for downstream analysis in case users do not read the documentation details (spoiler: most of them don't ;) )

Reply:
Ok. I understand your point and you are right, if CDR3 is not zero module 3 has no sense. So I guess is better to just not translate the entire nt sequence and mark the sequence as not translated. Because use a character as - may be considered as a gap and that can be confuse too.

Reply:
Sounds good to me. I was proposing a special character to find a middle ground between translating all the way and not translating. The idea was just to make an error pop in users downstream analysis due to unknown character. I agree `-` could be confusing with gaps. We could use another character that is not reserved or not translate at all, both seem good to me, your call!